### PR TITLE
Add support for propagating env values from *test rules

### LINKED ISSUE #906: Feature Request: Set runtime env vars for tests
Test targets can use the `env` attribute to supply environment variables for tests at runtime; however, these variables are not propagated when running the test target in BwB.

One potential solution is creating an `env` attribute on custom schemes for the test action similar to what we've done with the launch_action attributes. This would allow us to share the env var dict in the BUILD file with the test target and custom scheme.

## Tasks
- [x] #940 
- [x] #942 
- [x] Fix `Expand Variables Based On` setting when specifying test env vars/args
  - [x] #951 
  - [x] #956 
- [ ] Propagate env variables from `XXX_unit_test`. ([Request](https://github.com/buildbuddy-io/rules_xcodeproj/issues/906#issuecomment-1228751832))

### PR DESCRIPTION
**Pull request by maxwellE**
Fixes #906.

### CODE REVIEW

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 1]
```diff
@@ -256,6 +277,19 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
         ),
     )
 
+def _create_test_envs_depset(*, automatic_target_info, ctx, id, target):
+    test_env = getattr(ctx.rule.attr, automatic_target_info.env, {})
+    raw_run_env = target[RunEnvironmentInfo].environment if RunEnvironmentInfo in target else {}
+
+    # Some keys are not applicable in schemes, we will filter them out here
+    run_env = {}
+    denylist_run_env_keys = ["XCODE_VERSION_OVERRIDE", "XCODE_VERSION"]
```
Comment (maxwellE) [inline xcodeproj/internal/xcodeprojinfo.bzl:286]: @brentleyjones Not sure what is the right move here. Without this filtering we cannot test across multiple xcode versions at once since the will mutate the fixtures for each xcode version.

File: examples/integration/xcodeproj_targets.bzl [hunk 2]
```diff
@@ -47,6 +47,7 @@ XCODEPROJ_TARGETS = [
     "//tvOSApp/Test/UnitTests:tvOSAppUnitTests",
     "//watchOSApp/Test/UITests:watchOSAppUITests",
     "//watchOSAppExtension/Test/UnitTests:watchOSAppExtensionUnitTests",
+    "//iOSApp/Test/SwiftUnitTests:iOSAppSwiftUnitTests",
```
Comment (brentleyjones) [inline examples/integration/xcodeproj_targets.bzl:50]: Can you do this as it's own PR to keep the changes down?

File: examples/integration/test/fixtures/bwb.xcodeproj/xcshareddata/xcschemes/iOSAppSwiftUnitTests.xcscheme [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            isEnabled = "YES">
+         </EnvironmentVariable>
+         <EnvironmentVariable
+            key = "BUILD_WORKSPACE_DIRECTORY"
+            value = "$(BUILD_WORKSPACE_DIRECTORY)"
+            isEnabled = "YES">
+         </EnvironmentVariable>
+         <EnvironmentVariable
+            key = "IOSAPPSWIFTUNITTESTS"
+            value = "CUSTOM_ENV_VALUE"
```
Comment (brentleyjones) [inline examples/integration/test/fixtures/bwb.xcodeproj/xcshareddata/xcschemes/iOSAppSwiftUnitTests.xcscheme:115]: 🙌

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 4]
```diff
@@ -424,6 +464,8 @@ def create_xcodeprojinfo(*, ctx, target, transitive_infos):
 
     if _should_skip_target(ctx = ctx, target = target):
         info_fields = _skip_target(
+            automatic_target_info = automatic_target_info,
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:467]: Please move this to the bottom.

File: xcodeproj/internal/xcodeproj_rule.bzl [hunk 5]
```diff
@@ -873,6 +878,12 @@ def _xcodeproj_impl(ctx):
             transitive = [info.replacement_labels for info in infos],
         ).to_list()
     }
+    test_envs = {
+        r.id: r.env
+        for r in depset(
+            transitive = [info.test_envs for info in infos],
+        ).to_list()
+    }
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeproj_rule.bzl:886]:
```suggestion
    test_envs = {
        s.id: s.env
        for s in depset(
            transitive = [info.test_envs for info in infos],
        ).to_list()
        if s.env
    }
```

File: xcodeproj/internal/targets.bzl [hunk 6]
```diff
@@ -48,8 +48,6 @@ def _is_test_bundle(target, deps):
     if deps == None:
         return False
 
-    # TODO: Once we have a minimum version of Bazel 5.3+, we can check for
-    # `RunEnvironmentInfo` and `*XcTestBundleInfo` instead
```
Comment (brentleyjones) [inline xcodeproj/internal/targets.bzl:52]: Please put this comment back, since we haven't changed the check yet (though we can now).

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 7]
```diff
@@ -256,6 +277,19 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
...
+def _create_test_envs_depset(*, automatic_target_info, ctx, id, target):
+    test_env = getattr(ctx.rule.attr, automatic_target_info.env, {})
+    raw_run_env = target[RunEnvironmentInfo].environment if RunEnvironmentInfo in target else {}
+
+    # Some keys are not applicable in schemes, we will filter them out here
+    run_env = {}
+    denylist_run_env_keys = ["XCODE_VERSION_OVERRIDE", "XCODE_VERSION"]
+    for key, value in raw_run_env.items():
+        if key not in denylist_run_env_keys:
+            run_env[key] = value
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:289]:
```suggestion
    # Some keys are not applicable in schemes, we will filter them out here
    run_env = {
        key: value
        for key, value in raw_run_env.items()
        if key not in _IGNORED_BAZEL_ENV_VARIABLES ("XCODE_VERSION_OVERRIDE", "XCODE_VERSION")
    }
```

And then put this above the function:
```
_IGNORED_BAZEL_ENV_VARIABLES = {
    "XCODE_VERSION": None,
    "XCODE_VERSION_OVERRIDE": None,
}
```

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 8]
```diff
@@ -256,6 +277,19 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
...
+
+    return struct(id = id, env = struct(**dicts.add(test_env, run_env)))
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:291]: We want test_env to override run_env, right?
```suggestion
    return struct(id = id, env = struct(**dicts.add(run_env, test_env)))
```

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 9]
```diff
@@ -256,6 +277,19 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
         ),
     )
 
+def _create_test_envs_depset(*, automatic_target_info, ctx, id, target):
+    test_env = getattr(ctx.rule.attr, automatic_target_info.env, {})
+    raw_run_env = target[RunEnvironmentInfo].environment if RunEnvironmentInfo in target else {}
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:282]:
```suggestion
    raw_run_env = target[RunEnvironmentInfo].environment if RunEnvironmentInfo in target else {}
    test_env = getattr(ctx.rule.attr, automatic_target_info.env, {})
```

Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:282]: We need to update `README.md` to say 5.3 is the minimum Bazel version, unless we somehow detect the presence of a global symbol.

@keith @thii do either of you know how we could detect if the `RunEnvironmentInfo` provider is available? Does `hasattr(native, "RunEnvironmentInfo")` work?

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 10]
```diff
@@ -256,6 +277,19 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
         ),
     )
 
+def _create_test_envs_depset(*, automatic_target_info, ctx, id, target):
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:280]:
```suggestion
def _create_test_envs_depset(*, ctx, target, id, automatic_target_info):
```

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 11]
```diff
@@ -243,6 +251,19 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
         rule_kind = None,
         search_paths = search_paths,
         target_type = target_type.compile,
+        test_envs = depset(
+            [
+                _create_test_envs_depset(automatic_target_info = automatic_target_info, ctx = ctx, id = info.xcode_target.id, target = target)
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:256]:
```suggestion
                _create_test_envs_depset(
                    ctx = ctx,
                    target = target,
                    id = info.xcode_target.id,
                    automatic_target_info = automatic_target_info,
                )
```

File: xcodeproj/internal/providers.bzl [hunk 12]
```diff
@@ -33,6 +33,7 @@ attributes.
 An attribute name (or `None`) to collect `File`s from for the
 `entitlements`-like attribute.
 """,
+        "env": "env",
```
Comment (brentleyjones) [inline xcodeproj/internal/providers.bzl:36]: Expand the docs please.

**Review by brentleyjones — COMMENTED** [review]
I haven't looked at the generator code yet, just getting this part of the review in now.

File: xcodeproj/internal/providers.bzl [hunk 13]
```diff
@@ -150,6 +151,11 @@ target, it can still have a non-`None` value for this field.
         "transitive_dependencies": """\
 A `depset` of target ids (see the `target` `struct`) that this target
 transitively depends on.
+""",
+        "test_envs": """\
```
[the remark this answers is not in the document]
Comment (brentleyjones) [reply xcodeproj/internal/providers.bzl:155]: We will still need to have `envs` and `test_envs` in the spec.json though, since a scheme will need to set the correct ones for each type of target.

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 9 above)
[the remark this answers is not in the document]
Comment (tinder-maxwellelliott) [reply xcodeproj/internal/xcodeprojinfo.bzl:293]: Fixed the issue, should get --test_env now

File: examples/integration/xcodeproj_targets.bzl (hunk 2 above)
Comment (brentleyjones) [reply examples/integration/xcodeproj_targets.bzl]: Before merging I would like this change in another PR landed, so the fixture shows the `env` set.

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 14]
```diff
@@ -289,15 +289,19 @@ def _skip_target(
...
-def _create_test_envs_depset(*, ctx, target, id, automatic_target_info):
+def _create_envs_depset(*, ctx, target, id, automatic_target_info):
     raw_run_env = target[RunEnvironmentInfo].environment if RunEnvironmentInfo in target else {}
     test_env = getattr(ctx.rule.attr, automatic_target_info.env, {})
 
     # Some keys are not applicable in schemes, we will filter them out here
+    _IGNORED_BAZEL_ENV_VARIABLES = [
+        "XCODE_VERSION_OVERRIDE",
+        "XCODE_VERSION"
+    ]
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:300]: Let's put this outside of the function, so we don't recreate it each time it's called.

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 15]
```diff
@@ -256,6 +289,23 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
         ),
     )
 
+def _create_envs_depset(*, ctx, target, id, automatic_target_info):
+    raw_run_env = target[RunEnvironmentInfo].environment if RunEnvironmentInfo in target else {}
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:293]: If the `hasattr` check works, we need to use it here as well, otherwise this fails for Bazel < 5.3.

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 14 above)
Comment (tinder-maxwellelliott) [reply xcodeproj/internal/xcodeprojinfo.bzl:300]: Like in a constant in another bzl file?

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 16]
```diff
@@ -256,6 +289,25 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
...
+        "XCODE_VERSION_OVERRIDE",
+        "XCODE_VERSION",
+    ]
+    run_env = {
+        key: value
+        for key, value in raw_run_env.items()
+        if key not in _IGNORED_BAZEL_ENV_VARIABLES
+    }
+
+    return struct(id = id, env = struct(**dicts.add(run_env, test_env, ctx.configuration.test_env)))
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:309]:
```suggestion
    return struct(
        id = id,
        env = struct(
            **dicts.add(run_env, test_env, ctx.configuration.test_env)
        ),
    )
```

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 17]
```diff
@@ -256,6 +289,30 @@ def _skip_target(*, target, deps, deps_attrs, transitive_infos):
...
+    run_env = {
+        key: value
+        for key, value in raw_run_env.items()
+        if key not in _IGNORED_BAZEL_ENV_VARIABLES
+    }
+
+    return struct(
+        id = id,
+        env = struct(
+            **dicts.add(run_env, test_env, ctx.configuration.test_env)
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:312]: So `target[RunEnvironmentInfo].environment` didn't have `--test_env`? Does it have anything we want?

Comment (maxwellE) [reply xcodeproj/internal/xcodeprojinfo.bzl:314]: I dont think we actually need it for anything 🤞

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 18]
```diff
@@ -198,6 +198,12 @@ def _skip_target(
         transitive_infos = transitive_infos,
     )
 
+    # Some keys are not applicable in schemes, we will filter them out here
+    _IGNORED_BAZEL_ENV_VARIABLES = [
+        "XCODE_VERSION_OVERRIDE",
+        "XCODE_VERSION",
+    ]
+
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:206]: I'm probably not coming across correctly here, sorry. This should be a global constant in the file, not defined in any function, and referenced instead of passed. Also, a dictionary with `None` values for quick lookup.

File: xcodeproj/internal/targets.bzl (hunk 6 above)
Comment (brentleyjones) [reply xcodeproj/internal/targets.bzl:52]: Can we get the comment back?

File: xcodeproj/internal/xcodeproj_rule.bzl [hunk 19]
```diff
@@ -537,7 +538,8 @@ def _write_json_spec(
 "scheme_autogeneration_mode":"{scheme_autogeneration_mode}",\
 "target_hosts":{target_hosts},\
 "target_merges":{target_merges},\
-"targets":{targets}\
+"targets":{targets},\
+"envs": {envs}\
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeproj_rule.bzl:542]: Can you move this up to be alphabetical?

**Review by brentleyjones — COMMENTED** [review]
Can you move all of the `envs` to be alphabetical now?

**Comment by brentleyjones** [comment]
How does this handle if two tests in the same scheme set `envs`? I assume we can merge them if they don't conflict (right?), but I think we need to error out if they do. Can you add an example of this?

**Comment by maxwellE** [comment]
> How does this handle if two tests in the same scheme set `envs`? I assume we can merge them if they don't conflict (right?), but I think we need to error out if they do. Can you add an example of this?

Will add now

File: tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift [hunk 20]
```diff
@@ -101,8 +101,16 @@ Expected at least one target in `TestAction.targets`
         }
         for testActionTargetId in testActionTargetIds {
             if let testActionTargetEnv: [String: String] = envs[testActionTargetId] {
-                env.merge(testActionTargetEnv) { current, _ in
-                    current
+                for (key, value) in testActionTargetEnv {
+                    if let existingValue: String = env[key] {
+                        let errorMessage: String = """
+                        Detected conflicting environment variables:
```
Comment (maxwellE) [inline tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift:107]: @brentleyjones is this the right approach for conflicts?

Comment (brentleyjones) [reply tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift:107]: Should be a `UserError`, as `PreconditionError` is for bugs in the generator.

We only want to error if the values differ, because if they are the same they can merge just fine. I would use `.merging()` as you can throw in there by simply checking if the values differ.

File: tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift [hunk 21]
```diff
@@ -102,8 +102,11 @@ Expected at least one target in `TestAction.targets`
         for testActionTargetId in testActionTargetIds {
             if let testActionTargetEnv: [String: String] = envs[testActionTargetId] {
                 env.merging(testActionTargetEnv) { currentValue, incomingValue
-                    if currentValue != incomingValue {
-                        throw UsageError("ERROR Found environment variables with conflicting values ('\(currentValue)' and '\(incomingValue)')")
+                    guard currentValue == incomingValue else  {
+                        throw UsageError("""
+ERROR: Found environment variables with conflicting values ('\(currentValue)' \
```
[the remark this answers is not in the document]
Comment (maxwellE) [reply tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift:107]: Yea I like that, I realized that once we moved to merging. Will go back to that

File: examples/integration/xcodeproj_targets.bzl [hunk 22]
```diff
@@ -40,6 +40,7 @@ XCODEPROJ_TARGETS = [
         ],
         target_environments = ["device", "simulator"],
     ),
+    "//iOSApp/Test/ObjCUnitTests:iOSAppObjCUnitTests",
```
Comment (brentleyjones) [inline examples/integration/xcodeproj_targets.bzl:43]: Merge conflict?

File: tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift [hunk 23]
```diff
@@ -93,7 +94,31 @@ extension XCSchemeInfo.TestActionInfo {
...
+            return (targetId, label)
+        }
+        for tuple in testActionTargetIdsLabels {
+            let testActionTargetId: TargetID = tuple.0
+            let testActionLabel: BazelLabel = tuple.1
+            if let testActionTargetEnv: [String: String] = envs[testActionTargetId] {
+               for (key, newValue) in testActionTargetEnv {
+                    if let existingValue: String = env[key] {
+                        let errorMessage: String = """
+                        '\(testActionLabel)' defines a value for '\(key)' ('\(newValue)') that doesn't match the existing value of '\(existingValue)' from another target in the same scheme.
```
Comment (brentleyjones) [inline tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift:113]:
```suggestion
                        ERROR: '\(testActionLabel)' defines a value for '\(key)' ('\(newValue)') that doesn't match the existing value of '\(existingValue)' from another target in the same scheme.
```

File: tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift [hunk 24]
```diff
@@ -93,7 +94,31 @@ extension XCSchemeInfo.TestActionInfo {
...
+                        """
+                        throw UsageError(message: errorMessage)
```
Comment (brentleyjones) [inline tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift:115]: Did you test this locally in a custom scheme? Both the normal merge case and the conflict one? We should add the normal merge case into the fixture tests.

Comment (brentleyjones) [reply tools/generator/src/Generator/XCSchemeInfo+TestActionInfo.swift:115]: Was mainly saying that we should define one custom scheme in examples/integration that has both of the test targets in it to have the checked in fixture show merged env variables.

**Comment by maxwellE** [comment]
> Can you move all of the `envs` to be alphabetical now?

Like sort all functions parameters that use envs?

**Comment by brentleyjones** [comment]
Mostly. Just you had them in the spot that `test_env` was before you renamed it.

**Comment by maxwellE** [comment]
Modified the integration example to show merging of keys:

Screen Shot 2022-10-13 at 8 19 44 PM

**Review by brentleyjones — APPROVED** [review]
